gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java

```diff
@@ -0,0 +1,234 @@
+            Long.class,
+            BigInteger.class,
+            Float.class,
+            Double.class,
+            BigDecimal.class
+    };
+
+    private static final Comparator<Comparable> naturalOrderComparator = Comparator.naturalOrder();
+
+    /**
+     * This comparator does not provide a stable order for numerics because of type promotion equivalence semantics.
+     */
+    private static final Comparator<Number> numberComparator = (f,s) -> NumberHelper.compare(f, s);
+
+    /**
+     * This alternative numeric comparator will sort by numeric type in the case of type promotion equivalence.
+     * Currently unused.
+     */
+    private static final Comparator<Number> stableOrderNumberComparator = (f,s) -> {
+        final int i = NumberHelper.compare(f, s);
+        if (i != 0 || f.getClass().equals(s.getClass())) {
+            // they're not equal, or they are of the same numeric type
+            return i;
+        }
+        // if they are equal but of different types, order by type
+        final int ft = type(numerics, f);
+        final int st = type(numerics, s);
+        return ft - st;
+    };
+
+    /**
+     * Sort Vertex, Edge, VertexProperty by id.
+     */
+    private static final Comparator<Element> elementComparator =
+            Comparator.comparing(Element::id, INSTANCE);
+
+    /**
+     * Sort Property first by key, then by value.
+     */
+    private static final Comparator<Property> propertyComparator =
+            Comparator.<Property,Object>comparing(Property::key, INSTANCE).thenComparing(Property::value, INSTANCE);
+
+    /**
+     * Sort List, Set, Path, and Map element-by-element in the order presented by their natural iterator.
+     */
+    private static final Comparator<Iterable> iteratableComparator = (f, s) -> {
+        final Iterator fi = f.iterator();
+        final Iterator si = s.iterator();
+
+        while (fi.hasNext() && si.hasNext()) {
+            final int i = INSTANCE.compare(fi.next(), si.next());
+            if (i != 0) {
+                return i;
+            }
+        }
+
+        return fi.hasNext() ? 1 : si.hasNext() ? -1 : 0;
+    };
+
+    /**
+     * Sort Map by entry-set.
+     */
+    private static final Comparator<Map> mapComparator =
+            Comparator.comparing(Map::entrySet, iteratableComparator);
+
+    /**
+     * Sort Map.Entry first by key, then by value.
+     */
+    private static final Comparator<Map.Entry> entryComparator =
+            Comparator.<Map.Entry,Object>comparing(Map.Entry::getKey, INSTANCE).thenComparing(Map.Entry::getValue, INSTANCE);
+
+    /**
+     * Sort unknown types first by classname, then by natural toString().
+     */
+    private static final Comparator<Object> unknownTypeComparator =
+            Comparator.comparing(f -> f.getClass().getName()).thenComparing(Object::toString);
+
+    /**
+     * The comparators that will be used for the types defined above in the {@link #types} array. Be sure to keep them
```

Comment:
optional suggestion:
We can avoid potential bugs due to out of sync arrays by storing the mapping btw types and comparators in a LinkedHashMap<Class, Comparator>.